### PR TITLE
Add CloudTrail Rule to detect vulnerable EC2 AMIs re: CVE-2024-3094

### PR DESCRIPTION
**Pull request by egibs**
### Background

Detect the launch of AMIs that are listed as impacted by CVE-2024-3094.

Quick scripts to source the AMIs (Fedora only; OpenSUSE and Kali don't have affected [public] AMIs):

Bash:
```sh
#!/bin/bash

start_date="2024-03-26"
end_date="2024-04-02"

patterns=(
  "Fedora-Cloud-Base-AmazonEC2.x86_64-Rawhide*"
  "fedora-coreos-40*"
  "Fedora-Cloud-Base-AmazonEC2.aarch64-Rawhide*"
  "Fedora-Cloud-Base-40_Beta*"
)

regions=(
  "af-south-1"
  "ap-east-1"
  "ap-northeast-1"
  "ap-northeast-2"
  "ap-northeast-3"
  "ap-south-1"
  "ap-south-2"
  "ap-southeast-1"
  "ap-southeast-2"
  "ap-southeast-3"
  "ap-southeast-4"
  "ca-central-1"
  "ca-west-1"
  "eu-central-1"
  "eu-central-2"
  "eu-north-1"
  "eu-south-1"
  "eu-south-2"
  "eu-west-1"
  "eu-west-2"
  "eu-west-3"
  "il-central-1"
  "me-central-1"
  "me-south-1"
  "sa-east-1"
  "us-east-1"
  "us-east-2"
  "us-west-1"
  "us-west-2"
)

owner_id="125523088429"

amis=()

for region in "${regions[@]}"; do
  for pattern in "${patterns[@]}"; do
    
    ami_info=$(aws ec2 describe-images --region "$region" --owners "$owner_id" --filters "Name=name,Values=$pattern" --query 'Images[].{Id:ImageId,Name:Name,CreationDate:CreationDate}' --output json)

    if [ $? -ne 0 ]; then
      continue
    fi

    while IFS= read -r ami; do
      
      creation_date=$(echo "$ami" | jq -r '.CreationDate')
      formatted_date=$(date -j -f "%Y-%m-%dT%H:%M:%S.000Z" "$creation_date" +%Y-%m-%d 2>/dev/null)
      
      if [[ "$formatted_date" > "$start_date" && "$formatted_date" < "$end_date" ]]; then
        ami_id=$(echo "$ami" | jq -r '.Id')
        ami_name=$(echo "$ami" | jq -r '.Name')
        ami_info=$(echo "{\"Id\":\"$ami_id\",\"Name\":\"$ami_name\"}")
        amis+=("$ami_info")
      fi
    done <<< "$(echo "$ami_info" | jq -c '.[]')"
  done
done

if [ ${#amis[@]} -eq 0 ]; then
  echo "No AMIs found for the specified patterns and date range."
else
  json_output=$(echo "${amis[@]}" | jq -s '.')
  echo "$json_output"
fi
```

Python:
```py
import json
from datetime import datetime
import boto3
import botocore

# Set the start and end dates
start_date = "2024-03-26"
end_date = "2024-04-02"

# Define the AMI name patterns
patterns = {
    "Fedora-Cloud-Base-AmazonEC2.x86_64-Rawhide*",
    "fedora-coreos-40*",
    "Fedora-Cloud-Base-AmazonEC2.aarch64-Rawhide*",
    "Fedora-Cloud-Base-40_Beta*",
}

regions = {
  "af-south-1",
  "ap-east-1",
  "ap-northeast-1",
  "ap-northeast-2",
  "ap-northeast-3",
  "ap-south-1",
  "ap-south-2",
  "ap-southeast-1",
  "ap-southeast-2",
  "ap-southeast-3",
  "ap-southeast-4",
  "ca-central-1",
  "ca-west-1",
  "eu-central-1",
  "eu-central-2",
  "eu-north-1",
  "eu-south-1",
  "eu-south-2",
  "eu-west-1",
  "eu-west-2",
  "eu-west-3",
  "il-central-1",
  "me-central-1",
  "me-south-1",
  "sa-east-1",
  "us-east-1",
  "us-east-2",
  "us-west-1",
  "us-west-2",
}

# Set the owner account ID
owner_id = "125523088429"

# Initialize an empty list to store the AMI information
amis = []

# Iterate over the AMI name patterns
for region in regions:
    ec2_client = boto3.client(region_name=region, service_name="ec2")
    for pattern in patterns:
        # Get the AMI information based on the pattern and owner account ID
        try:
            response = ec2_client.describe_images(
                Owners=[owner_id],
                Filters=[
                    {
                        "Name": "name",
                        "Values": [pattern]
                    }
                ]
            )

            # Iterate over each AMI
            for ami in response["Images"]:
                # Extract the creation date of the AMI
                creation_date = ami["CreationDate"]

                # Convert the creation date to the format YYYY-MM-DD
                formatted_date = datetime.strptime(creation_date, "%Y-%m-%dT%H:%M:%S.%fZ").strftime("%Y-%m-%d")

                # Compare the formatted date with the start and end dates
                if start_date < formatted_date < end_date:
                    # Extract the AMI ID and Name
                    ami_id = ami["ImageId"]
                    ami_name = ami["Name"]

                    # Create a dictionary with the AMI ID and Name
                    ami_info = {
                        "Id": ami_id,
                        "Name": ami_name,
                        "Region": region,
                    }

                    # Append the AMI information to the list
                    amis.append(ami_info)
        except botocore.exceptions.ClientError:
            continue

# Check if the AMIs list is empty
if not amis:
    print("No AMIs found for the specified patterns and date range.")
else:
    # Convert the list of AMI information to JSON format
    # sort the amis by region
    amis.sort(key=lambda x: x["Region"])
    json_output = json.dumps(amis, indent=2)

    # Output the JSON result
    print(json_output)
```

Red Hat's guidance:
```
PLEASE IMMEDIATELY STOP USAGE OF ANY FEDORA RAWHIDE INSTANCES for work or personal activity. Fedora Rawhide will be reverted to xz-5.4.x shortly, and once that is done, Fedora Rawhide instances can safely be redeployed. Note that Fedora Rawhide is the development distribution of Fedora Linux, and serves as the basis for future Fedora Linux builds (in this case, the yet-to-be-released Fedora Linux 41).

At this time the Fedora Linux 40 builds have not been shown to be compromised. We believe the malicious code injection did not take effect in these builds. However, Fedora Linux 40 users should still downgrade to a 5.4 build to be safe. An update that reverts xz to 5.4.x has recently been published and is becoming available to Fedora Linux 40 users through the normal update system. Concerned users can force the update by following the instructions at https://bodhi.fedoraproject.org/updates/FEDORA-2024-d02c7bb266.
```

### Changes

- Adds `AWS.EC2.Vulnerable.XZ.Image.Launched`

### Testing

`make fmt; make lint; make test`

### CODE REVIEW

**Review by arielkr256 — APPROVED** [review]
Very nice!  Too bad we can't put the image ID scripts in the repo.

File: rules/aws_cloudtrail_rules/aws_ec2_vulnerable_xz_image_launched.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+from panther_base_helpers import aws_rule_context
+from panther_default import aws_cloudtrail_success
+
+# AMIs published by Fedora between 2022-03-26 and 2022-04-02
+# OpenSUSE and Kali do not have any recent [public] AMIs that would be affected
+
+# pylint: disable=line-too-long
+AMIS = {
```
Comment (jacknagz) [inline rules/aws_cloudtrail_rules/aws_ec2_vulnerable_xz_image_launched.py:8]: Should we put this into a LUT instead given that AMIs are region-specific?

Comment (egibs) [reply rules/aws_cloudtrail_rules/aws_ec2_vulnerable_xz_image_launched.py:8]: Ah, yea, probably. I didn't even think about that.

File: global_helpers/panther_aws_ami.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+AnalysisType: global
+Filename: panther_aws_ami.py
```
Comment (jacknagz) [inline global_helpers/panther_aws_ami.yml:2]: Do we need a new helper? Could we use an existing one? If not, can we rename this to be more specific that it's about vuln-xz-aws-amis?

**Review by jacknagz — APPROVED** [review]

**Review by arielkr256 — APPROVED** [review]

File: global_helpers/panther_aws_ami.yml (hunk 2 above)
Comment (egibs) [reply global_helpers/panther_aws_ami.yml:2]: `panther_iocs` may be a good fit for it, actually. I'll refactor this.

Comment (egibs) [reply global_helpers/panther_aws_ami.yml]: Changed in [`c78dd0c` (#1182)](https://github.com/panther-labs/panther-analysis/pull/1182/commits/c78dd0c744c5121ace4781433931269389d36cdf).

**Pull request merged**